core/model/src/main/java/org/eclipse/rdf4j/model/util/GraphComparisons.java

```diff
@@ -0,0 +1,447 @@
+					}
+					return automorphism;
+				}
+			}
+		}
+		return null;
+	}
+
+	protected static List<Collection<BNode>> partitions(Multimap<HashCode, BNode> partitionMapping) {
+		List<Collection<BNode>> partition = new ArrayList<>();
+
+		for (Entry<HashCode, Collection<BNode>> entry : partitionMapping.asMap().entrySet()) {
+			partition.add(entry.getValue());
+		}
+		return partition;
+	}
+
+	protected static Multimap<HashCode, BNode> partitionMapping(Map<BNode, HashCode> blankNodeMapping) {
+		return Multimaps.invertFrom(Multimaps.forMap(blankNodeMapping), HashMultimap.create());
+	}
+
+	private static BigInteger mappingSize(Map<BNode, HashCode> mapping) {
+		BigInteger size = mapping.values()
+				.stream()
+				.map(h -> new BigInteger(1, h.asBytes()))
+				.reduce(BigInteger.ZERO, (v1, v2) -> v1.add(v2));
+		return size;
+	}
+
+	private static Model labelModel(Model original, Map<BNode, HashCode> hash) {
+		Model result = new DynamicModelFactory().createEmptyModel();
+
+		for (Statement st : original) {
+			if (st.getSubject().isBNode() || st.getObject().isBNode()
+					|| (st.getContext() != null && st.getContext().isBNode())) {
+				Resource subject = st.getSubject().isBNode()
+						? createCanonicalBNode((BNode) st.getSubject(), hash)
+						: st.getSubject();
+				IRI predicate = st.getPredicate();
+				Value object = st.getObject().isBNode()
+						? createCanonicalBNode((BNode) st.getObject(), hash)
+						: st.getObject();
+				Resource context = (st.getContext() != null && st.getContext().isBNode())
+						? createCanonicalBNode((BNode) st.getContext(), hash)
+						: st.getContext();
+
+				result.add(subject, predicate, object, context);
+			} else {
+				result.add(st);
+			}
+		}
+		return result;
+	}
+
+	protected static Map<BNode, HashCode> hashBNodes(Model m) {
+		return hashBNodes(m, null);
+	}
+
+	private static Map<BNode, HashCode> hashBNodes(Model m, Map<BNode, HashCode> initialBlankNodeMapping) {
+		final Map<Value, HashCode> staticValueMapping = new HashMap<>();
+
+		final Set<BNode> blankNodes = getBlankNodes(m);
+
+		final Map<BNode, HashCode> initialHash = initialBlankNodeMapping == null ? new HashMap<>(blankNodes.size())
+				: initialBlankNodeMapping;
+
+		if (initialHash.isEmpty()) {
+			blankNodes.forEach(node -> initialHash.put(node, initialHashCode));
+		}
+
+		Map<BNode, HashCode> currentHash = null;
+		if (blankNodes.isEmpty()) {
+			currentHash = initialHash;
+		} else {
+			Map<BNode, HashCode> previousHash = initialHash;
+			do {
+				Map<BNode, HashCode> temp = currentHash;
+				currentHash = new HashMap<>(previousHash);
+				previousHash = temp != null ? temp : initialHash;
+
+				for (BNode b : blankNodes) {
+					for (Statement st : m.getStatements(b, null, null)) {
+						HashCode c = hashTuple(hashForValue(st.getObject(), previousHash, staticValueMapping),
+								hashForValue(st.getPredicate(), previousHash, staticValueMapping), outgoing);
+						currentHash.put(b, hashBag(c, currentHash.get(b)));
```

Comment:
If BNodes could hold their own hashes, then this line would be something like `b.addHash(hashBag(c, b.getHashVersion(version)))`.

Reply:
I'm considering something along these lines as a test, but it will require quite a bit of rewiring and is unlikely to be done before your milestone build is due.